qiime2/core/annotate.py

```diff
@@ -351,3 +395,319 @@ def _write(self, annotations_dir, root_result_uuid,
 
         with open(note_path, 'wb') as fh:
             fh.write(contents)
+
+
+# public key algorithm identifiers
+# https://datatracker.ietf.org/doc/html/rfc4880#section-9.1
+_PUBKEY_ALG = {
+    '1': 'RSA',
+    '2': 'RSA',
+    '3': 'RSA',
+    '16': 'ElGamal',
+    '17': 'DSA',
+    '18': 'ECDH',
+    '19': 'ECDSA',
+    '22': 'EdDSA'
+}
+
+
+# SIGNATURE HELPERS
+# helper for locating root_fp for a given Result
+def _find_root_fp(start):
+    p = pathlib.Path(start).resolve()
+
+    for _ in range(6):
+        provenance = p / 'provenance'
+        meta_yaml = p / 'metadata.yaml'
+        if provenance.is_dir() and meta_yaml.is_file():
+            return p
+        elif p.parent == p:
+            break
+        p = p.parent
+    raise ValueError(f'Could not locate Result root starting from {start}.')
```

Comment:
What you wrote here is something that would probably go in util, but we don't need it in most cases because USUALLY if we need this information, we can access `Result._archiver.path` which maddeningly enough you're one step too far removed to do here. For this specific use case though, I would probably do something like the following.

```suggestion
def _find_root_fp(annotations_dir, root_result_uuid):
    split_fp = annotations_dir.split(os.sep)
    root_result_uuid_index = split_fp.index(root_result_uuid)
    root_fp = os.sep.join(split_fp[0: root_result_uuid_index + 1])
    return root_path
```

We know the root path we are looking for is above the `annotations_dir` we have, and we know it terminates with the `root_result_uuid`, so we can leverage this to split the path we have apart and recombine it only up to the point we need. The way you did it would work even if we didn't have the root_uuid, but in this case we do.

We also theoretically in this case do know EXACTLY where we are relative to the root of the artifact we are annotating, and so we could just hop up one level if we want to assume the structure won't change.

Reply:
this makes sense! and yeah, it was maddeningly just out of reach.